src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -3427,4 +3503,78 @@ mod tests {
             assert_eq!(got, exp);
         }
     }
+
+    #[test]
+    fn test_insert_binary() {
+        let cases = vec![
+            ("abc", 0, -1, "def", "abc"),
+            ("abc", 0, 0, "def", "abc"),
+            ("abc", 0, 3, "def", "abc"),
+            ("abc", 0, 4, "def", "abc"),
+            ("abc", 1, -1, "def", "def"),
+            ("abc", 1, 0, "def", "defabc"),
+            ("abc", 1, 1, "def", "defbc"),
+            ("abc", 1, 3, "def", "def"),
+            ("abc", 1, 4, "def", "def"),
+            ("abc", 2, -1, "def", "adef"),
+            ("abc", 2, 0, "def", "adefbc"),
+            ("abc", 2, 1, "def", "adefc"),
+            ("abc", 2, 2, "def", "adef"),
+            ("abc", 2, 3, "def", "adef"),
+            ("abc", 3, -1, "def", "abdef"),
+            ("abc", 3, 0, "def", "abdefc"),
+            ("abc", 3, 1, "def", "abdef"),
+            ("abc", 3, 2, "def", "abdef"),
+        ];
+        for (s, pos, len, new_str, expected) in cases {
+            let got: Datum = eval_func(
+                ScalarFuncSig::InsertBinary,
+                &[
+                    Datum::Bytes(s.as_bytes().to_vec()),
+                    Datum::I64(pos),
+                    Datum::I64(len),
+                    Datum::Bytes(new_str.as_bytes().to_vec()),
+                ],
+            )
+            .unwrap();
+            assert_eq!(got, Datum::Bytes(expected.as_bytes().to_vec()));
+        }
+    }
+    #[test]
+    fn test_insert() {
+        let cases = vec![
+            ("你好，", 0, -1, "世界", "你好，"),
+            ("你好，", 0, 0, "世界", "你好，"),
+            ("你好，", 0, 3, "世界", "你好，"),
+            ("你好，", 0, 4, "世界", "你好，"),
+            ("你好，", 1, -1, "世界", "世界"),
+            ("你好，", 1, 0, "世界", "世界你好，"),
+            ("你好，", 1, 1, "世界", "世界好，"),
+            ("你好，", 1, 3, "世界", "世界"),
+            ("你好，", 1, 4, "世界", "世界"),
+            ("你好，", 2, -1, "世界", "你世界"),
+            ("你好，", 2, 0, "世界", "你世界好，"),
+            ("你好，", 2, 1, "世界", "你世界，"),
+            ("你好，", 2, 2, "世界", "你世界"),
+            ("你好，", 2, 3, "世界", "你世界"),
+            ("你好，", 3, -1, "世界", "你好世界"),
+            ("你好，", 3, 0, "世界", "你好世界，"),
+            ("你好，", 3, 1, "世界", "你好世界"),
+            ("你好，", 3, 2, "世界", "你好世界"),
+        ];
+        for (s, pos, len, new_str, expected) in cases {
```

Comment:
Can you also test if one of the arguments is `NULL`?

Reply:
@rleungx Updated.